### ISSUE TITLE
Implement prs-7 request/response handling

### ISSUE THREAD

**Issue by Hikariii** [onelogin/php-saml#162]
At the moment this library does a lot of request parsing and response handling on it's own.
This issue proposes moving to [psr-7](http://www.php-fig.org/psr/psr-7/) for request/response handling.

### Bypass and/or resolve current and possible future issues:
- Assumptions are made about the incoming request and superglobals are being interpreted sometimes leading to misbehaving code.
    - https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Utils.php#L363-L382
    - https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Utils.php#L402-L435
    - https://github.com/onelogin/php-saml/issues/149#issue-175214993
- The toolkit (in some cases) sets headers and exits (in case of redirecting). It is possible to circumvent, but it is considered bad practise to even have the option of exiting php within a library.
    - https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Utils.php#L270-L279

### There is a standard, please use it:
Up until last year here was a lot of different request/response handling going on within different frameworks. Since psr-7 a lot of frameworks have moved to prs-7 for their response/request handling.
Just check https://mwop.net/blog/2015-01-26-psr-7-by-example.html and how this is a great standard for handling this.

### Increase interoperability and collaboration:
Using psr-7 internally would greatly improve the integration in different frameworks and would make it a lot easier to write middleware around php-saml. Requests and responses could be interpreted and returned in a standardized psr-7 way, clear for everyone who uses it.

### Provide backwards compatibility layer for old-style usage:
To provide a form of backwards compatibility the library could provide it's own utility to create and resolve psr-7 requests and responses for usage outside of a framework. I actually forsee that most of the implementations will actually use this library within a framework and do not need this layer, but it's still nice to have and be able to use this library in a stand-alone way.

### Requires a lot of refactoring
This is probably something for a 3.* version, because it will change the outer api.
Internally it will probably be quite some work to upgrade this, but it will definitely improve adoption and ease of use for this library within the context of bigger applications.

**Comment by pitbulk** [comment]
prs-7 seems interesting and something to take in mind. Thanks or sharing this.
Maybe if people collaborate with can release that in a 3.* version.

**Comment by chris001** [comment]
@Hikariii @pitbulk 
> This is probably something for a 3.* version, because it will change the outer api.
It should probably be done by adding another interface which consumes and produces psr7 type format parameters, while keeping the old pre-psr7 interface which uses urls and parameters in arrays and strings.  Then internally the new code can handle the data objects however is best. Probably best to add a class which does the conversion for you, and can output its contents either in psr7 format, or scheme/uri/url/parameters/cookies/etc strings format. Probably this class already exists.

**Comment by pitbulk** [comment]
I think a 3.X version makes sense. Focused on PHP7 and PRE-PSR7 and [witthout the mcrypt requirement](https://github.com/onelogin/php-saml/tree/remove_mcrypt)

**Comment by pitbulk** [comment]
The 3.X is now there, but we need to keep improving it in order to do an official release.
@Hikariii @chris001 do you have any progress related to port php-saml to use PSR-7?

**Comment by chris001** [comment]
@pitbulk 
To accomplish this I suggest adding:
* composer and psr-4 autoloading and obviously namespace all the php source files into its own namespace to prevent collisions in the global namespace,
* guzzlehttp psr-7 it's the most popular library to implement psr-7 https://github.com/guzzle/psr7 it's easy to use and easy to install at first run with a composer command.
```
git clone https://github.com/onelogin/php-saml
cd php-saml
composer require guzzlehttp/psr7 guzzlehttp/guzzle
```

**Comment by Hikariii** [comment]
@chris001 and @pitbulk: I started an experimental rewrite a few weeks ago. Hope to finish a great deal of it this week.
For PSR-7 I use Zend Diactoros https://github.com/zendframework/zend-diactoros.
Guzzle is a http-client app and the PSR-7 Requests/Responses for SAML are for a php server-side app.

**Comment by Hikariii** [comment]
Also, I'm working with php-saml in a Symfony app and would like to use the Symfony PSR-7 bridge with php-saml to inject requests and return responses.
The PSR-7 bridge already suggests working with Zend Diactoros: https://symfony.com/doc/current/request/psr7.html

**Comment by pitbulk** [comment]
Hi @Hikariii, any progress on this?

**Comment by pitbulk** [comment]
@Hikariii I made some progress on the 3.0.0 branch. I want to know if you have time to work on it, or if I may continue your work of that branch: https://github.com/Hikariii/php-saml/commits/feature-rewrite

**Comment by stof** [comment]
What is the status of this work ? I would be interested in this, as it would integrate much more cleanly in my Symfony project.

**Comment by pitbulk** [comment]
Any help is welcome, I'm currently out of time since I had to jump in other projects.

**Comment by slaughter550** [comment]
@pitbulk Did you like the idea of using https://github.com/zendframework/zend-diactoros or would you be open to using https://github.com/guzzle/psr7? I have substantially more experience with guzzle but happy to learn more about zend-diactoros if thats the direction you want to go.

**Comment by pitbulk** [comment]
Based on @Hikariii I think makes more sense to use zend-diactoros

**Comment by slaughter550** [comment]
Sounds good, I'll review what @Hikariii did and keep pushing forward

**Comment by stof** [comment]
Reviewing what @Hikariii has done is almost impossible, as he deleted the whole codebase and added the new code in a different folder (making any git diffing unusable)

**Comment by Hikariii** [comment]
@stof I agree that I did do far too many changes to the codebase to come to a PR that is diffable. The work I did ended up to be a restructure/rewrite combined with integrating new features(too much going on).
I think the way forward must include a lot more smaller steps. The hard part about that is that taking a request/response object as input/output is inherently different from the way the codebase works now.

The codebase heavily uses php globals and gets it input from $_GET and $_POST. Even more, the code also writes its own headers and even exits (can be bypassed).
I explored a way to really use a request/response object as a base, but concluded that it came down to a lot of rewriting.

I do think it is wise to explore a way of restructuring the code in a way that it can be reviewed and refactored nicely and in smaller bits.

**Comment by Jorijn** [comment]
@Hikariii is this still on your radar somewhere?

**Comment by Hikariii** [comment]
@Jorijn Nope not really. Refactoring this in a structured way requires a lot of planning/testing/rewriting and probably some help. It is not a priority (anymore).

**Comment by joelharkes** [comment]
Doesn't seem to much work to simply make the $_GET variable an optional input parameter?

https://github.com/search?q=repo%3ASAML-Toolkits%2Fphp-saml%20%24_GET&type=code

Its only in Auth and LogoutRequest?
